src/test/lib/ApiTest.cpp

```diff
@@ -4521,6 +4521,188 @@ void QuicTest_QUIC_PARAM_CONN_ORIG_DEST_CID(MsQuicRegistration& Registration, Ms
+            MsQuicConnection Connection(Registration);
+            TEST_QUIC_SUCCEEDED(Connection.GetInitStatus());
+            uint16_t Port = 4433;
+
+            for (uint8_t i = 0; i < 4; i++) {
+                QUIC_ADDR ClientAddr;
+                QuicAddrFromString("127.0.0.1", Port++, &ClientAddr);
+                TEST_QUIC_SUCCEEDED(
+                    Connection.SetParam(
+                        QUIC_PARAM_CONN_ADD_LOCAL_ADDRESS,
+                        sizeof(ClientAddr),
+                        &ClientAddr));
+            }
+        }
+    }
+}
+
+void QuicTest_QUIC_PARAM_CONN_REMOVE_LOCAL_ADDRESS(MsQuicRegistration& Registration, MsQuicConfiguration& ClientConfiguration)
+{
+    TestScopeLogger LogScope0("QUIC_PARAM_CONN_REMOVE_LOCAL_ADDRESS");
+    //
+    // SetParam
+    //
+    {
+        TestScopeLogger LogScope1("SetParam");
+        //
+        // No local address to remove
+        //
+        {
+            TestScopeLogger LogScope2("No local address to remove");
+            TEST_TRUE(ClientConfiguration.IsValid());
+            MsQuicConnection Connection(Registration);
+            TEST_QUIC_SUCCEEDED(Connection.GetInitStatus());
+
+            QUIC_ADDR Dummy = {};
+            TEST_QUIC_STATUS(
+                QUIC_STATUS_NOT_FOUND,
+                Connection.SetParam(
+                    QUIC_PARAM_CONN_REMOVE_LOCAL_ADDRESS,
+                    sizeof(Dummy),
+                    &Dummy));
+        }
+
+        //
+        // Add and remove a local address
+        //
+        {
+            TestScopeLogger LogScope2("Add and remove a local address");
+            MsQuicConnection Connection(Registration);
+            TEST_QUIC_SUCCEEDED(Connection.GetInitStatus());
+            QUIC_ADDR Dummy = {};
+            TEST_QUIC_SUCCEEDED(
+                Connection.SetParam(
+                    QUIC_PARAM_CONN_ADD_LOCAL_ADDRESS,
+                    sizeof(Dummy),
+                    &Dummy));
+
+            TEST_QUIC_SUCCEEDED(
+                Connection.SetParam(
+                    QUIC_PARAM_CONN_REMOVE_LOCAL_ADDRESS,
+                    sizeof(Dummy),
+                    &Dummy));
+        }
+
+        //
+        // Remove a local address that belongs to another the active path
+        //
+        {
+            TestScopeLogger LogScope2("Remove a local address that belongs to another the active path");
+            MsQuicConnection Connection(Registration);
+            TEST_QUIC_SUCCEEDED(Connection.GetInitStatus());
+            TEST_QUIC_SUCCEEDED(
+                MsQuic->ConnectionStart(
+                    Connection.Handle,
+                    ClientConfiguration,
+                    QUIC_ADDRESS_FAMILY_INET,
+                    "localhost",
+                    4433));
+
+            QuicAddr ClientLocalAddr;
+            TEST_QUIC_SUCCEEDED(Connection.GetLocalAddr(ClientLocalAddr));
+            TEST_QUIC_STATUS(
+                QUIC_STATUS_INVALID_STATE,
```

Comment:
Can/should we really block removal of current active path? In the future, removal can/should be in response to the address being removed from the system itself (i.e. Wi-Fi gets disconnected). You can't say 'no, you're not allowed because I'm using it'. So, really, I think we need to allow removal, but if that's the last one, the connection just dies.

Reply:
I agree with you. How should we handle the challenge? Separate PR?